test/unit/asrouter/CFRPageActions.test.js

```diff
@@ -0,0 +1,634 @@
+        sandbox.spy(pageAction, "_clearScheduledStateChanges");
+        sandbox.spy(pageAction, "_expand");
+        sandbox.spy(pageAction, "_dispatchImpression");
+
+        await pageAction.show(fakeRecommendation);
+        assert.notCalled(pageAction._clearScheduledStateChanges);
+        assert.notCalled(pageAction._expand);
+        assert.notCalled(pageAction._dispatchImpression);
+
+        await pageAction.show(fakeRecommendation, true);
+        assert.calledOnce(pageAction._clearScheduledStateChanges);
+        assert.calledOnce(pageAction._expand);
+        assert.calledWith(pageAction._expand, 1000);
+        assert.calledOnce(pageAction._dispatchImpression);
+        assert.calledWith(pageAction._dispatchImpression, fakeRecommendation);
+      });
+      it("should send telemetry if `expand` is true and the id and bucket_id are provided", async () => {
+        sandbox.spy(pageAction, "_sendTelemetry");
+
+        await pageAction.show(fakeRecommendation, true);
+        const expectedParam = {
+          message_id: fakeRecommendation.id,
+          bucket_id: fakeRecommendation.content.bucket_id,
+          event: "IMPRESSION"
+        };
+        assert.calledOnce(pageAction._sendTelemetry);
+        assert.calledWith(pageAction._sendTelemetry, expectedParam);
+      });
+    });
+
+    describe("#hide", () => {
+      it("should hide the container, cancel any state changes, and remove the state attribute", () => {
+        sandbox.spy(pageAction, "_clearScheduledStateChanges");
+        pageAction.hide();
+        assert.isTrue(pageAction.container.hidden);
+        assert.calledOnce(pageAction._clearScheduledStateChanges);
+        assert.notOk(pageAction.urlbar.getAttribute("cfr-recommendation-state"));
+      });
+      it("should remove the `currentNotification`", () => {
+        const notification = {};
+        pageAction.currentNotification = notification;
+        pageAction.hide();
+        assert.calledWith(global.PopupNotifications.remove, notification);
+      });
+    });
+
+    describe("#_expand", () => {
+      beforeEach(() => {
+        pageAction._clearScheduledStateChanges();
+        pageAction.urlbar.removeAttribute("cfr-recommendation-state");
+      });
+      it("without a delay, should clear other state changes and set the state to 'expanded'", () => {
+        sandbox.spy(pageAction, "_clearScheduledStateChanges");
+        pageAction._expand();
+        assert.calledOnce(pageAction._clearScheduledStateChanges);
+        assert.equal(pageAction.urlbar.getAttribute("cfr-recommendation-state"), "expanded");
+      });
+      it("with a delay, should set the expanded state after the correct amount of time", () => {
+        const delay = 1234;
+        pageAction._expand(delay);
+        assert.lengthOf(pageAction.stateTransitionTimeoutIDs, 1);
+        assert.notOk(pageAction.urlbar.getAttribute("cfr-recommendation-state"));
+        clock.tick(delay - 1);
```

Comment:
I think these assertions are too specific: testing the correct behaviour of setTimeout. I would instead only test that the correct attribute is set. It could be grouped with the test for the `#show` method and not have an explicit call for `_expand`.
I don't see any use for the `delay = 0` branch. Can that be removed?

Reply:
@AdamHillier I meant https://github.com/mozilla/activity-stream/blob/f0443a23070f3b68b13546c40c5e881b3e2921d5/lib/CFRPageActions.jsm#L124-L128